pkg/controller/machine/controller.go

```diff
@@ -154,6 +154,16 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 			return reconcile.Result{}, err
 		}
 
+		if m.Status.NodeRef != nil {
+			klog.Infof("Deleting node %q", m.Status.NodeRef.Name)
+			var node corev1.Node
+			if err := r.Client.Get(ctx, client.ObjectKey{Name: m.Status.NodeRef.Name}, &node); err != nil {
+				klog.Warningf("Failed to get node %q for machine %q: %v", m.Status.NodeRef.Name, m.Name, err)
+			} else if err := r.Client.Delete(ctx, &node); err != nil {
```

Comment:
Why is this in an  `else` branch? IMHO it should:

* Return an error if ` r.Client.Get(ctx, client.ObjectKey{Name: m.Status.NodeRef.Name}` returns an error that is not kubernetesapierrors.IsNotExist
* Return nil if getting the node returns a `kubernetesapierrors.IsNotExist` err
* Proceed to delete the node if there was no error when getting the node
* Return an error and not just log a warning if an attempt to delete the node was made and did not succeed

Reply:
I would probably factor the whole node deletion out to not require so much branching in the code, the `Return nil if getting the node returns a kubernetesapierrors.IsNotExist err` would only work if the node deletion happened in a distinct func

Reply:
thanks! the reason for being loose with the node deletion errors initially is that if we requeue on error, the machine will go through r.actuator.Delete again which might fail the second run (as the instance does not exist now) depending on the actuator implementation.
wdyt? If we agree this is fine I'll push the commit addressing your comments

Reply:
With the approach I suggested we always check if the node exist, if we get an error that is not "not found" we return that error and req the item, if we get an "not found" error we don't delete, if we get no error we do delete and if that deletion failes we req again.

The machine will still exist as the object has a finalizer that gets removed after the node cleanup code. Providers should be able to handle the case that an instance for a machine with a non-nil deletion timestamp doesnt exist

Reply:
yep,  my concern is about providers not handle it. Updated, thanks!